### PR TITLE
refactor: Simplify duplicated simulation and improve test condition

### PR DESCRIPTION
**Pull request by jollidah**
1. Simplify duplicated situation(leaving leader) with for loop
2. Develop test conditions:
	- Before code change: 
		1. Validate leader_id
		2. Validate cluster size
	- After code change:
		1. Validate leader_id
		2. Validate cluster size
		3. Validate follower’s term
		4. Track raw node's StateRole change

### CODE REVIEW

File: harness/src/utils.rs [hunk 1]
```diff
@@ -128,6 +130,22 @@ pub fn kill_process_using_port(port: u16) {
...
+            .iter()
+            .filter(|(_, role)| role.role == InitialRole::Voter)
+            .for_each(|(id, _)| {
+                hash_set.insert(*id);
+            });
+    }
+    hash_set.len()
+}
+
 pub fn kill_previous_raft_processes() {
```
Comment (jopemachine) [inline harness/src/utils.rs:149]: Is there a reason why the counting needs to be executed multiple times for 1 second, rather than just once after a 1-second delay?

File: harness/src/utils.rs [hunk 2]
```diff
@@ -128,6 +130,22 @@ pub fn kill_process_using_port(port: u16) {
...
+            .await
+            .inner
+            .iter()
+            .filter(|(_, role)| role.role == InitialRole::Voter)
+            .for_each(|(id, _)| {
+                hash_set.insert(*id);
+            });
+    }
+    hash_set.len()
+}
```
Comment (jopemachine) [inline harness/src/utils.rs:147]: Is there a reason why the counting needs to be executed multiple times for 1 second, rather than just once after a 1-second delay?

Comment (jollidah) [reply harness/src/utils.rs:147]: I use a loop because the tick interval is unpredictable, so repeatedly checking within the 2-second window ensures we capture all relevant state changes.

File: harness/tests/bootstrap.rs [hunk 3]
```diff
@@ -18,10 +18,6 @@ pub async fn test_static_bootstrap() {
 
     sleep(Duration::from_secs(1)).await;
 
-    let raft_1 = rafts.get(&1).unwrap();
-
-    wait_for_until_cluster_size_increase(raft_1.clone(), 3).await;
```
Comment (jopemachine) [inline harness/tests/bootstrap.rs:23]: Why is it necessary to remove `wait_for_until_cluster_size_increase`?
This test is meant to ensure that nodes 1, 2, and 3 are all functioning properly.
(This has been replaced with waiting until the cluster size reaches 3 to confirm that all three nodes are functioning properly.)

So, if `wait_for_until_cluster_size_increase` is removed, this test code will lose its meaning.

File: harness/tests/bootstrap.rs [hunk 4]
```diff
@@ -59,10 +55,6 @@ pub async fn test_dynamic_bootstrap() {
 
     let mut rafts = wait_until_rafts_ready(None, rx_raft, 3).await;
 
-    let raft_1 = rafts.get(&1).unwrap();
-
-    wait_for_until_cluster_size_increase(raft_1.clone(), 3).await;
```
Comment (jopemachine) [inline harness/tests/bootstrap.rs:64]: Same here

File: harness/src/utils.rs (hunk 2 above)
Comment (jopemachine) [reply harness/src/utils.rs:132]: `tick_interval` is fixed.
If you are referring to `election_tick`, we can fix it using `min_election_tick` and `max_election_tick`.

File: harness/tests/leader_election.rs [hunk 5]
```diff
@@ -1,120 +1,68 @@
...
-    wait_for_until_cluster_size_decrease(raft_k.clone(), 3).await;
-    sleep(Duration::from_secs(2)).await;
-
-    let leader_id = raft_k.get_leader_id().await;
-
-    assert!(leader_id != 0);
-    assert_eq!(raft_k.get_cluster_size().await, 3);
-
-    sleep(Duration::from_secs(2)).await;
+    run_leader_election_test(FIVE_NODE_EXAMPLE, vec![1, 2, 3, 4, 5]).await;
```
Comment (jopemachine) [inline harness/tests/leader_election.rs:62]: Since the list of `node_id`s already exists in the example file, passing a separate list as an argument would be redundant.

File: harness/tests/leader_election.rs [hunk 6]
```diff
@@ -1,120 +1,68 @@
...
-    assert!(
-        [2, 3].contains(&leader_id),
-        "Actual leader_id: {}",
-        leader_id
-    );
+    let mut current_node_cnt = raft_list.len();
+
+    let mut rafts = wait_until_rafts_ready(None, rx_raft, raft_list.len()).await;
+    let mut ex_term = 1;
+    for _ in 1..3 {
```
Comment (jopemachine) [inline harness/tests/leader_election.rs:22]: Instead of performing it twice arbitrarily, how about passing the number of iterations as an argument?

File: harness/src/utils.rs [hunk 7]
```diff
@@ -128,8 +93,81 @@ pub fn kill_process_using_port(port: u16) {
         .expect("Failed to execute kill command");
 }
 
+pub async fn collect_state_matching_rafts_until_leader_emerge(
```
Comment (jopemachine) [inline harness/src/utils.rs:96]: I think `collect_state_matching_rafts_until_leader_emerge` is verbose and performs too many tasks.
I think it would be better to separate the functionality of returning rafts that match a specific state and waiting for a leader to emerge.

File: harness/src/utils.rs [hunk 8]
```diff
@@ -128,8 +93,81 @@ pub fn kill_process_using_port(port: u16) {
...
+    rafts: &HashMap<u64, Raft>,
+    target_role: StateRole,
+) -> HashSet<u64> {
+    let mut target_set = HashSet::with_capacity(rafts.len());
+    let mut leader_merged = false;
```
Comment (jopemachine) [inline harness/src/utils.rs:101]:
```suggestion
    let mut leader_merged = false;
```

File: harness/src/utils.rs [hunk 9]
```diff
@@ -128,8 +93,81 @@ pub fn kill_process_using_port(port: u16) {
...
+    let mut ex_term = 1;
+    // ! This loop can cause infinite hang
```
Comment (jopemachine) [inline harness/src/utils.rs:103]: Comments like these would be better placed as function's documentation rather than inline within the code.

File: raftify/src/raft_node/mod.rs [hunk 10]
```diff
@@ -1147,7 +1147,10 @@ impl<
                 tx_msg
                     .send(LocalResponseMsg::GetRawNode {
                         raw_node: Arc::new(Mutex::new(unsafe {
-                            std::mem::transmute(&self.raw_node)
+                            std::mem::transmute::<
+                                &jopemachine_raft::RawNode<HeedStorage>,
+                                &jopemachine_raft::RawNode<HeedStorage>,
```
Comment (jopemachine) [inline raftify/src/raft_node/mod.rs:1152]: Could you explain why this change is necessary?

File: harness/src/utils.rs (hunk 2 above)
Comment (jollidah) [reply harness/src/utils.rs:132]: Currently, tick_interval is fixed in this test. We need to modify it to handle varying tick_interval values

File: harness/tests/bootstrap.rs (hunk 3 above)
Comment (jollidah) [reply harness/tests/bootstrap.rs:23]: I think the wait_until_rafts_ready function would ensure the cluster size increases

File: harness/tests/leader_election.rs (hunk 6 above)
Comment (jollidah) [reply harness/tests/leader_election.rs:22]: That'll be much vetter

File: raftify/src/raft_node/mod.rs (hunk 10 above)
Comment (jollidah) [reply raftify/src/raft_node/mod.rs:1152]: To clarify the namespace, but i think it's unnecessary

File: harness/tests/leader_election.rs (hunk 5 above)
Comment (jollidah) [reply harness/tests/leader_election.rs:62]: Thanks! I'll fix it

File: harness/src/utils.rs (hunk 9 above)
Comment (jollidah) [reply harness/src/utils.rs:103]: Thanks!
